[GEOT-7632]  WMS capabilities parsing should not fail if 'units' for a dimension is empty

skalesse opened the pull request:
[![GEOT-7632](https://badgen.net/badge/JIRA/GEOT-7632/0052CC)](https://osgeo-org.atlassian.net/browse/GEOT-7632) [<img width="16" alt="Powered by Pull Request Badge" src="https://user-images.githubusercontent.com/1393946/111216524-d2bb8e00-85d4-11eb-821b-ed4c00989c02.png">](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

This PR is to fix [[GEOT-7632]  WMS capabilities parsing should not fail if 'units' for a dimension is empty](https://osgeo-org.atlassian.net/browse/GEOT-7632). According to [OpenGIS Web Map Service WMS Implementation Specification](https://portal.ogc.org/files/?artifact_id=14416), for layer dimensions, `units` must **not** be missing, **but are allowed to be empty**:

> If the dimensional quantity has no units (e.g. band number in a multi-wavelength sensor), use the null string: units="".

`gt-wms` currently throws a `SAXException` in case `units.length == 0`. The two affected classes are `org.geotools.ows.wms.xml.WMSComplexTypes._DimensionType` and `org.geotools.ows.wms.xml.Dimension#Dimension`.

This PR fixes the problem by modifying the two above mentioned places to remove the length check while parsing the `units` attribute. Two unit tests have been added for some basic dimension value checks (it looks no such test-cases ever existed before). These checks include the described case. 

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

<!--Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.-->

aaime commented on modules/extension/wms/src/test/java/org/geotools/ows/wms/xml/test/WMSComplexTypesTest.java:
```diff
@@ -171,4 +172,127 @@ public void testEmptyOnlineResourceServiceDef() throws Exception {
+        Assert.assertNotNull(dimension);
+        // unit
+        units = dimension.getUnits();
+        Assert.assertNotNull("Units must not be missing", units);
+        Assert.assertEquals("Unit should be empty .", "", units);
+        // unit symbol
+        Assert.assertEquals("Unit symbol should be 'm'", "m", dimension.getUnitSymbol());
```
The example seems a bit odd but ok, it's just a test.
According to the specification:

>  The elevation units identifier “verticalCRSid” refers to a vertical CRS as in 6.7.5. The unitSymbol attribute shall be chosen to match the units of the specified vertical CRS. 

An example of unit-less dimension is a band number, quoting the relevant section from page 52 (WMS 1.3 spec):

>   * If the dimensional quantity has no units (e.g. band number in a multi-wavelength sensor), use the null string: units="". 
>   * If the dimensional quantity has units, unit names should be taken from the Unified Code for Units of Measure (UCUM) if UCUM has an appropriate entry. When UCUM is used, the mandatory units attribute shall be an appropriate entry from the UCUM “name” column.
>  * If present, the unitSymbol shall be a 7-bit ASCII character string. When UCUM is used, the unitSymbol shall be the entry from the UCUM “c/s” column (case-sensitive abbreviation) corresponding to the UCUM name.  UCUM prefix names and symbols may be present as well.

skalesse replied:
Yes you are right, the test is an odd example, but I just didn't want to make it more complicated. It still should show what we are aiming for, I guess.

aaime commented on modules/extension/wms/src/main/java/org/geotools/ows/wms/xml/Dimension.java:
```diff
@@ -51,7 +51,8 @@ public Dimension(String name, String units, String unitSymbol) {
         }
-        if (units == null || units.length() == 0) {
+        // unit must not be null, but can be empty
+        if (units == null) {
```
This is a bizzarre choice in the spec, I would quote the relevant spec section in the class javadoc for clarification (to clarify the code is not just leveraging a loophole in the schema, it was the actual specification intention):

> * If the dimensional quantity has no units (e.g. band number in a multi-wavelength sensor), use the null string: units="".

aaime reviewed (changes requested):
Looks good, I would just clarify that it was really the specification intent in the Dimension.java javadoc.

skalesse commented:
My original intention had been to leave as much code the way it was, but given yours (and copilot's) comments, I have made the following two requested changes: 

- I have added the relevant section to the class JavaDoc of `Dimension.java`. Note that I have also cleaned up as much as possible the JavaDoc, because the code blocks and citations were formatted incorrectly. Note that I was not able to remove the empty lines in the block-quote. The code fomatter always brings them back. 
- I have fixed copilot's complains about the typos and wrong error messages. Note that while I was at it I have also removed the code duplication and fixed a minor complaint about the use of `String#length() == 0` vs. `String#isEmpty()`.

Let me know what you think. 
Thanks

aaime reviewed (approved):
Looks good to me
